tests/systemtests/authz_test.go

```diff
@@ -0,0 +1,840 @@
+	}{
+		{
+			"valid transaction",
+			granteeAddr,
+			val1Addr,
+			10,
+			false,
+			"",
+		},
+		{
+			"undelegate to not allowed address",
+			granteeAddr,
+			val2Addr,
+			10,
+			true,
+			"cannot delegate/undelegate",
+		},
+	}
+
+	for _, tc := range undelegateTestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			undelegateTx := fmt.Sprintf(`{"@type":"%s","delegator_address":"%s","validator_address":"%s","amount":{"denom":"%s","amount":"%d"}}`,
+				msgUndelegateTypeURL, granterAddr, tc.valAddr, denom, tc.amount)
+			execMsg := WriteToTempJSONFile(t, undelegateTx)
+			defer execMsg.Close()
+
+			cmd := append(append(execCmdArgs, execMsg.Name()), "--from="+tc.grantee)
+			if tc.expectErr {
+				rsp := cli.Run(cmd...)
+				RequireTxFailure(t, rsp)
+				require.Contains(t, rsp, tc.expErrMsg)
+			} else {
+				rsp := cli.RunAndWait(cmd...)
+				RequireTxSuccess(t, rsp)
+
+				// query delegation and check balance reduced
+				expectedAmount := delegatedAmount - tc.amount
+				resp = cli.CustomQuery("q", "staking", "delegation", granterAddr, val1Addr)
+				delegatedAmount = gjson.Get(resp, "delegation_response.balance.amount").Int()
+				require.Equal(t, expectedAmount, delegatedAmount)
+			}
+		})
+	}
+
+	// revoke existing grant
+	rsp = cli.RunAndWait("tx", "authz", "revoke", granteeAddr, msgUndelegateTypeURL, "--from", granterAddr)
+	RequireTxSuccess(t, rsp)
+
+	// check grants between granter and grantee after revoking
+	resp = cli.CustomQuery("q", "authz", "grants", granterAddr, granteeAddr)
+	grants = gjson.Get(resp, "grants").Array()
+	require.Len(t, grants, 0)
+
+	// test exec redelegate authorization
+	rsp = cli.RunAndWait("tx", "authz", "grant", granteeAddr, "redelegate",
+		fmt.Sprintf("--allowed-validators=%s,%s", val1Addr, val2Addr),
+		"--fees=1stake",
+		"--from", granterAddr)
+	RequireTxSuccess(t, rsp)
+
+	var redelegationAmount int64 = 10
+
+	redelegateTx := fmt.Sprintf(`{"@type":"%s","delegator_address":"%s","validator_src_address":"%s","validator_dst_address":"%s","amount":{"denom":"%s","amount":"%d"}}`,
+		msgRedelegateTypeURL, granterAddr, val1Addr, val2Addr, denom, redelegationAmount)
+	execMsg = WriteToTempJSONFile(t, redelegateTx)
+	defer execMsg.Close()
+
+	redelegateCmd := append(append(execCmdArgs, execMsg.Name()), "--from="+granteeAddr, "--gas=auto")
+	rsp = cli.RunAndWait(redelegateCmd...)
+	RequireTxSuccess(t, rsp)
+
+	// query new delegation and check balance increased
+	resp = cli.CustomQuery("q", "staking", "delegation", granterAddr, val2Addr)
+	delegatedAmount = gjson.Get(resp, "delegation_response.balance.amount").Int()
+	require.GreaterOrEqual(t, delegatedAmount, redelegationAmount)
+
+	// revoke all existing grants
+	rsp = cli.RunAndWait("tx", "authz", "revoke-all", "--from", granterAddr)
+	RequireTxSuccess(t, rsp)
+
+	// check grants after revoking
+	resp = cli.CustomQuery("q", "authz", "grants-by-granter", granterAddr)
+	grants = gjson.Get(resp, "grants").Array()
+	require.Len(t, grants, 0)
+}
```

Comment:
This was a very big test. Although I do understand that all the cases share the exec command, do you see some way to simplify this or split into multiple tests? You are covering send, generic, delegate, undelegate
as well as use cases like expired and and revoked. I would expect most of this is covered by unit tests already.

Reply:
I am trying to migrate tests already present in tests/e2e. So, I had implemented all those which were present. Will try to split them into multiple tests.